crates/connector-init/src/capture.rs

```diff
@@ -15,39 +16,65 @@ impl proto_grpc::capture::driver_server::Driver for Driver {
         &self,
         request: tonic::Request<SpecRequest>,
     ) -> Result<tonic::Response<SpecResponse>, tonic::Status> {
-        let message = rpc::unary(&self.entrypoint, "spec", request.into_inner()).await?;
+        let message =
+            rpc::unary(&self.entrypoint, self.codec, "spec", request.into_inner()).await?;
         Ok(tonic::Response::new(message))
     }
 
     async fn discover(
         &self,
         request: tonic::Request<DiscoverRequest>,
     ) -> Result<tonic::Response<DiscoverResponse>, tonic::Status> {
-        let message = rpc::unary(&self.entrypoint, "discover", request.into_inner()).await?;
+        let message = rpc::unary(
+            &self.entrypoint,
+            self.codec,
+            "discover",
+            request.into_inner(),
+        )
+        .await?;
         Ok(tonic::Response::new(message))
     }
 
     async fn validate(
         &self,
         request: tonic::Request<ValidateRequest>,
     ) -> Result<tonic::Response<ValidateResponse>, tonic::Status> {
-        let message = rpc::unary(&self.entrypoint, "validate", request.into_inner()).await?;
+        let message = rpc::unary(
+            &self.entrypoint,
+            self.codec,
+            "validate",
+            request.into_inner(),
+        )
+        .await?;
         Ok(tonic::Response::new(message))
     }
 
     async fn apply_upsert(
         &self,
         request: tonic::Request<ApplyRequest>,
     ) -> Result<tonic::Response<ApplyResponse>, tonic::Status> {
-        let message = rpc::unary(&self.entrypoint, "apply-upsert", request.into_inner()).await?;
+        let message = rpc::unary(
+            &self.entrypoint,
+            self.codec,
+            "apply-upsert",
+            request.into_inner(),
+        )
+        .await?;
         Ok(tonic::Response::new(message))
     }
 
     async fn apply_delete(
         &self,
         request: tonic::Request<ApplyRequest>,
     ) -> Result<tonic::Response<ApplyResponse>, tonic::Status> {
-        let message = rpc::unary(&self.entrypoint, "apply-delete", request.into_inner()).await?;
+        // For the JSON protocol, there is no apply-delete operation.
+        // Instead, a deletion is an apply with no bindings.
+        let mut request = request.into_inner();
+        if let Codec::Json = self.codec {
+            request.capture.as_mut().unwrap().bindings.clear();
+        }
```

Comment:
Is this a change we want to extend to non-JSON protocol as well? If not, I would prefer if we don't diverge the JSON protocol from the protobuf protocol 🤔

Reply:
Yes, the proposal is that we _remove_ deletion as an RPC of the protobuf protocol, and instead make deletion an application with empty bindings.

This is addressing a tension we discussed during the offsite, where connectors are weird because they drop tables / resources if you delete the whole materialization, but not if you delete bindings of the ongoing materialization. The intention is to consolidate into a single thing: from the connector's perspective (*not* the UI or control-plane), the materialization / capture is never "deleted" but it will be asked to apply updates to the bindings over time, and those bindings could be empty.

It's safer to do it first with the JSON protocol, since that's entirely new, and we can update protobuf connectors over time. And, if the JSON protocol is working well for us, I'd actually propose we make it a goal to deprecate and remove the use of protobuf altogether in our connectors (in which case this migration is defacto accomplished), but let's take it one step at a time.